File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-go/Mailstandard/provider_selection.go
package Mailstandard

import "strings"

type ParsedMailProviderUrl struct {
    ProviderKey string
    RawUrl      string
}

type MailProviderSelection struct {
    ProviderKey string
    Source      string
}

type MailProviderSelectionRequest struct {
    ProviderUrl                  string
    ProviderKey                  string
    TenantOverrideProviderKey    string
    DeploymentProfileProviderKey string
}

var MailProviderSelectionSources = []string{
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
}

var MailProviderSelectionPrecedence = []string{
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
}

func hasMailProviderSelectionText(value string) bool {
    return strings.TrimSpace(value) != ""
}

func ParseMailProviderUrl(providerUrl string) ParsedMailProviderUrl {
    trimmed := strings.TrimSpace(providerUrl)
    if !strings.HasPrefix(trimmed, "Mail:") || !strings.Contains(trimmed, "://") {
        panic("Invalid Mail provider URL: " + providerUrl)
    }

    withoutPrefix := strings.TrimPrefix(trimmed, "Mail:")
    providerKey, _, _ := strings.Cut(withoutPrefix, "://")

    return ParsedMailProviderUrl{
        ProviderKey: strings.ToLower(providerKey),
        RawUrl:      providerUrl,
    }
}

func ResolveMailProviderSelection(
    request MailProviderSelectionRequest,
    defaultProviderKey string,
) MailProviderSelection {
    if hasMailProviderSelectionText(request.ProviderUrl) {
        return MailProviderSelection{
            ProviderKey: ParseMailProviderUrl(request.ProviderUrl).ProviderKey,
            Source:      "provider_url",
        }
    }

    if hasMailProviderSelectionText(request.ProviderKey) {
        return MailProviderSelection{
            ProviderKey: strings.TrimSpace(request.ProviderKey),
            Source:      "provider_key",
        }
    }

    if hasMailProviderSelectionText(request.TenantOverrideProviderKey) {
        return MailProviderSelection{
            ProviderKey: strings.TrimSpace(request.TenantOverrideProviderKey),
            Source:      "tenant_override",
        }
    }

    if hasMailProviderSelectionText(request.DeploymentProfileProviderKey) {
        return MailProviderSelection{
            ProviderKey: strings.TrimSpace(request.DeploymentProfileProviderKey),
            Source:      "deployment_profile",
        }
    }

    resolvedDefaultProviderKey := defaultProviderKey
    if !hasMailProviderSelectionText(resolvedDefaultProviderKey) {
        resolvedDefaultProviderKey = DEFAULT_mail_PROVIDER_KEY
    }

    return MailProviderSelection{
        ProviderKey: resolvedDefaultProviderKey,
        Source:      "default_provider",
    }
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-typescript/src/provider-selection.ts
import { MailSdkException } from './errors.js';
import { DEFAULT_mail_PROVIDER_KEY } from './provider-catalog.js';
import { freezeMailRuntimeValue } from './runtime-freeze.js';
import type {
  MailProviderSelection,
  MailProviderSelectionRequest,
  MailProviderSelectionSource,
} from './types.js';

const mail_PROVIDER_URL_PATTERN = /^Mail:([a-z0-9-]+):\/\/.+$/i;

export interface ParsedMailProviderUrl {
  providerKey: string;
  rawUrl: string;
}

export const mail_PROVIDER_SELECTION_SOURCES = freezeMailRuntimeValue([
  'provider_url',
  'provider_key',
  'tenant_override',
  'deployment_profile',
  'default_provider',
] as const);

export const mail_PROVIDER_SELECTION_PRECEDENCE = freezeMailRuntimeValue([
  ...mail_PROVIDER_SELECTION_SOURCES,
]);

export function parseMailProviderUrl(providerUrl: string): ParsedMailProviderUrl {
  const match = mail_PROVIDER_URL_PATTERN.exec(providerUrl.trim());
  if (!match) {
    throw new MailSdkException({
      code: 'invalid_provider_url',
      message: `Invalid Mail provider URL: ${providerUrl}`,
      details: { providerUrl },
    });
  }
  const providerKey = match[1];
  if (!providerKey) {
    throw new MailSdkException({
      code: 'invalid_provider_url',
      message: `Invalid Mail provider URL: ${providerUrl}`,
      details: { providerUrl },
    });
  }

  return freezeMailRuntimeValue({
    providerKey: providerKey.toLowerCase(),
    rawUrl: providerUrl,
  });
}

function createMailProviderSelection(
  providerKey: string,
  source: MailProviderSelectionSource,
): MailProviderSelection {
  return freezeMailRuntimeValue({
    providerKey,
    source,
  });
}

export function resolveMailProviderSelection(
  request: MailProviderSelectionRequest = {},
  defaultProviderKey: string = DEFAULT_mail_PROVIDER_KEY,
): MailProviderSelection {
  if (request.providerUrl) {
    return createMailProviderSelection(
      parseMailProviderUrl(request.providerUrl).providerKey,
      'provider_url',
    );
  }

  if (request.providerKey) {
    return createMailProviderSelection(request.providerKey, 'provider_key');
  }

  if (request.tenantOverrideProviderKey) {
    return createMailProviderSelection(
      request.tenantOverrideProviderKey,
      'tenant_override',
    );
  }

  if (request.deploymentProfileProviderKey) {
    return createMailProviderSelection(
      request.deploymentProfileProviderKey,
      'deployment_profile',
    );
  }

  return createMailProviderSelection(defaultProviderKey, 'default_provider');
}
